pep-0646.rst

```diff
@@ -733,28 +733,166 @@ hopefully enable a thriving ecosystem of tools for analysing and verifying
+        def __iter__(self):
+            yield self._unpacked
 
-1. Star expressions must be made valid in at least index operations.
-   For example, ``Tuple[*Ts]`` and ``Tuple[T1, *Ts, T2]`` would both
-   be valid. (This PEP does not allow multiple unpacked ``TypeVarTuple``
-   instances to appear in a single parameter list, so ``Tuple[*Ts1, *Ts2]``
-   would be a runtime error. Also note that star expressions would *not*
-   be valid in slice expressions -  e.g. ``Tuple[*Ts:*Ts]`` is
-   nonsensical and should remain invalid.)
-2. We would need to make '``*args: *Ts``' valid in function definitions.
+There are three things to note about this grammar change. First, for consistency with
+star-unpacking in other contexts, ``_UnpackedTypeVarTuple`` would always be received
+by ``__class_getitem__`` as part of a tuple:
+
+::
+
+    class C:
+        def __class_getitem__(self, x):
+            ...
+            
+    [*(1,)]      # = [1]
+    [*(1,), 2]   # = [1, 2]
+    C[int]       # __class_getitem__ receives x=int
+    # But perhaps surprisingly:
+    C[*Ts]       # __class_getitem__ receives x=(Ts._unpacked, )
+    # This is consistent with:
+    C[*Ts, int]  # __class_getitem__ receives x=(Ts._unpacked, int)
+
+
+Second, the change would also enable star-unpacking of other structures within
+indexing operations:
+
+::
+
+    idxs_to_select = (1, 2, 3)
+    array[0, *idxs_to_select, -1]
+
+Third, more than one instance of a star-unpack can occur within an index:
+
+::
+
+    array[*idxs_to_select, *idxs_to_select]
+
+The requirement that only one unpacked TypeVarTuple may be used per list of type parameters
+would therefore be implemented only in type checking tools themselves rather than at the
+syntax level.
+
+Finally, note that star expressions within slices remain invalid:
+
+::
 
-In both cases, at runtime the star operator would call ``Ts.__iter__()``.
-This would, in turn, return an instance of a helper class, e.g.
-``UnpackedTypeVarTuple``, whose ``repr`` would be ``*Ts``.
+    # Syntax error
+    array[*idxs_start:*idxs_end]
 
-If these grammar changes are considered too burdensome, we could instead
-simply use ``Unpack`` - though in this case it might be better for us to
-first decide whether there's a better option.
+``*args`` as a TypeVarTuple
+---------------------------
+
+The second change enables use of ``*args: *Ts``.
+
+Before:
+
+::
+
+    star_etc[StarEtc*]:
+    | '*' a=param_no_default b=param_maybe_default* c=[kwds] {
+        _PyPegen_star_etc(p, a, b, c) }
+    | '*' ',' b=param_maybe_default+ c=[kwds] {
+        _PyPegen_star_etc(p, NULL, b, c) }
+    | a=kwds { _PyPegen_star_etc(p, NULL, NULL, a) }
+
+After:
+
+::
+
+    star_etc[StarEtc*]:
+    | '*' a=param_no_default b=param_maybe_default* c=[kwds] {
+        _PyPegen_star_etc(p, a, b, c) }
+    | '*' a=param_no_default_star_annotation b=param_maybe_default* c=[kwds] {
+        _PyPegen_star_etc(p, a, b, c) }
```

Comment:
Maybe point out that this is the new alternative, whereas the two other alternatives are unchanged.

Reply:
Done.